File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        if not l1 or not l2:
            return l1 if l1 else l2 
        dummy = ListNode(0)
        tail = dummy 
        carry = 0 
        while l1 or l2 or carry:
            val = 0 
            if l1:
                val+=l1.val
                l1 = l1.next
            if l2:
                val+=l2.val
                l2 = l2.next 
            val+=carry
            carry = val//10 
            val=val%10 
            node = ListNode(val)
            tail.next = node
            tail = tail.next
        return dummy.next

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode * dummy = new ListNode(0);
        ListNode * tail = dummy;
        ListNode *curr1 = l1;
        ListNode *curr2 = l2;
        int carry = 0 ;
        int val1=0;
        int val2=0;
        int val=0;
        while(curr1 || curr2 || carry)
        {
            if(curr1)
            {val1 = curr1->val;
            curr1=curr1->next;
            }
            else
            val1 = 0 ;

            if(curr2)
            {val2 = curr2->val;
            curr2=curr2->next;}
            else
            val2=0;

            

            val = val1+val2+carry;
            carry = val/10;
            val=val%10;
            tail->next = new ListNode(val);
            tail = tail->next;

        }
        
        return dummy->next;
    }
};